Vector tiles polylines

bagnell opened the pull request:
For https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/25.

This is a branch off `vector-tiles-polygons` so I opened a PR into that to minimized the code diff. Either wait until both this and #4186 are ready or I can open a new PR into `vector-tiles`.

* [ ] Add unit tests
* [ ] Add doc
* [ ] Investigate polygon offset values
* [ ] Investigate incomplete polyline rendering issues (more likely an issue with the tileset)
* [ ] Add per feature colors
* [ ] Render during `GROUND` pass?
* [ ] Update 3D Tiles Sandcastle example (could be just one vector example)

lilleyse commented on Source/Scene/Cesium3DTileGroundPolylines.js:
```diff
@@ -0,0 +1,418 @@
+        PolylineColorAppearanceVS,
+        BlendingState,
+        Pass
+    ) {
```
Formatting

```javascript
        Pass) {
    'use strict';
```

lilleyse commented on Source/Scene/Cesium3DTileGroundPolylines.js:
```diff
@@ -0,0 +1,418 @@
+        var indices = IndexDatatype.createTypedArray(size, positionsLength * 6 - 6);
+        var index = 0;
+        var indicesIndex = 0;
+        length = positionsLength - 1.0;
```
`- 1;`

lilleyse commented on Source/Scene/Cesium3DTileGroundPolylines.js:
```diff
@@ -0,0 +1,418 @@
+            polygonOffset : polygonOffset
+        });
+
+        primitive._rsPick = RenderState.fromCache({
```
This variable isn't in the constructor.

lilleyse commented on Source/Scene/Cesium3DTileGroundPolylines.js:
```diff
@@ -0,0 +1,418 @@
+            defines : ['VECTOR_TILE'],
+            sources : [fsSource]
+        });
+        primitive._spPick = ShaderProgram.fromCache({
```
Same here, and it's not handled in destroy.

lilleyse commented on Source/Scene/Vector3DTileContent.js:
```diff
@@ -272,38 +280,76 @@ define([
-            boundingVolume : this._tile._boundingVolume.boundingVolume,
-            batchTableResources : this.batchTableResources
-        });
+        if (positions.length > 0 && false) {
```
Is this area still in progress?

lilleyse commented:
Looks good so far.

How easy would it be to support dynamic width?

lilleyse commented:
I set the width to 10 and am seeing some artifacts. Is this already considered as part of your to-do list?
![capture](https://cloud.githubusercontent.com/assets/915398/17755102/04fb9964-64a6-11e6-929e-51cc1f198cac.JPG)

pjcozzi commented:
Added to tasklist:

> * [ ] Update 3D Tiles Sandcastle example (could be just one vector example)

pjcozzi commented on Source/Scene/Cesium3DTileGroundPolylines.js:
```diff
@@ -0,0 +1,418 @@
+/*global define*/
+define([
```
Should we rename this class to something more generic in case it is used for other batched renderers?

pjcozzi replied:
This organization would be similar to instanced models.

pjcozzi commented on Source/Scene/Cesium3DTileGroundPolylines.js:
```diff
@@ -0,0 +1,418 @@
+    'use strict';
+
+    function Cesium3DTileGroundPolylines(options) {
+        this._positions = options.positions;
```
Remind me, do we need to keep these in system memory?  Can we just keep indices to support dynamic styling?

bagnell replied:
None of these are kept in memory. We will need widths for styling unless we can generalize the batch table to include it.

pjcozzi commented:
>  Render during GROUND pass?

As discussed offline at one point, I think it is fine to have `GROUND_OPAQUE` and `GROUND_TRANSLUCENT` passes.

pjcozzi commented on Source/Scene/Cesium3DTileGroundPolylines.js:
```diff
@@ -0,0 +1,418 @@
+                shaderProgram : primitive._sp,
+                uniformMap : uniformMap,
+                boundingVolume : primitive._boundingVolume,
+                modelMatrix : Matrix4.IDENTITY,
```
Isn't this the default?

pjcozzi commented on Source/Scene/Cesium3DTileGroundPolylines.js:
```diff
@@ -0,0 +1,418 @@
+                shaderProgram : primitive._spPick,
+                uniformMap : uniformMap,
+                boundingVolume : primitive._boundingVolume,
+                modelMatrix : Matrix4.IDENTITY,
```
Same comment.

pjcozzi commented on Source/Shaders/Appearances/PolylineColorAppearanceVS.glsl:
```diff
@@ -1,3 +1,12 @@
+#ifdef VECTOR_TILE
```
Is this uber shader better than having two separate shaders and perhaps shared GLSL functions as needed?

pjcozzi commented:
Just those comments.

bagnell commented:
>>  Render during GROUND pass?

>As discussed offline at one point, I think it is fine to have `GROUND_OPAQUE` and `GROUND_TRANSLUCENT` passes.

As part of this PR? Or in a separate PR to master?

bagnell commented:
>How easy would it be to support dynamic width?

It should be pretty easy. I want to add it to the batch table instead of having it as a vertex attribute.

bagnell commented:
I'm going to merge this into the `vector-tiles-polygons` branch because its easier to work on them both in one branch. I've addressed most of the comments here. I'll move any that are still open and the checklist to the other PR.

pjcozzi commented:
>>> Render during GROUND pass?

>> As discussed offline at one point, I think it is fine to have GROUND_OPAQUE and GROUND_TRANSLUCENT passes.

> As part of this PR? Or in a separate PR to master?

If this needs to be done for vector tiles, I would like to have this PR go into master so it doesn't still in `3d-tiles` waiting for the spec.
